File: Data Structures & Algorithms/stone-game-ii/submission-3.java
class Solution {
    private int[] suf_sum;
    private int[][] memo;

    public int stoneGameII(int[] piles) {
        int n = piles.length;
        suf_sum = new int[n + 1];
        memo = new int[n][n + 1];

        for (int i = n - 1; i >= 0; i--) suf_sum[i] = suf_sum[i + 1] + piles[i];

        return dp(0, 1, n);
    }

    private int dp(int i, int m, int n) {
        if ((i + 2 * m) >= n)
            return suf_sum[i];

        if (memo[i][m] != 0)
            return memo[i][m];
        int bst = 0;

        for (int x = 1; x < 2 * m + 1; x++) {
            int opp = dp(i + x, Math.max(m, x), n);
            int cur = suf_sum[i] - opp;
            bst = Math.max(cur, bst);
        }

        memo[i][m] = bst;
        return bst;
    }
}

File: Data Structures & Algorithms/stone-game/submission-3.py
class Solution:
    def stoneGame(self, piles: List[int]) -> bool:
        n = len(piles)
        dp = [0] * n

        for l in range(n - 1, -1, -1):
            for r in range(l, n):
                if (r - l) % 2 == 0:
                    lt = piles[l]
                    rt = piles[r]
                else:
                    lt = rt = 0

                if l == r:
                    dp[r] = rt
                else:
                    dp[r] = max(dp[r] + lt, dp[r - 1] + rt)

        tot = 0
        for pile in piles:
            tot += pile
        a = dp[n - 1]

        return a > (tot - a)
